php/php.editor/src/org/netbeans/modules/php/editor/completion/PHPCodeCompletion.java

```diff
@@ -1073,7 +1061,61 @@ private void autoCompleteInClassContext(
                     }
                 }
             }
+        } else if (completeFieldTypes(tokenSequence, caretOffset, th, info.getSnapshot().getSource().getFileObject())){
+            autoCompleteFieldType(completionResult, request);
+        }
+    }
+
+    private void autoCompleteFieldType(final PHPCompletionResult completionResult, CompletionRequest request) {
+        // PHP 7.4 Typed Properties 2.0
+        autoCompleteNamespaces(completionResult, request);
+        autoCompleteTypeNames(completionResult, request);
+        autoCompleteKeywords(completionResult, request, Type.getTypesForFieldType());
+    }
+
+    private boolean offerMagicAndInherited(TokenSequence<PHPTokenId> tokenSequence, int caretOffset, TokenHierarchy<?> th) {
+        boolean offerMagicAndInherited = true;
+        tokenSequence.move(caretOffset);
+        if (!(!tokenSequence.moveNext() && !tokenSequence.movePrevious())) {
+            Token<PHPTokenId> token = tokenSequence.token();
+            int tokenIdOffset = tokenSequence.token().offset(th);
+            offerMagicAndInherited = !CompletionContextFinder.lineContainsAny(token, caretOffset - tokenIdOffset, tokenSequence, Arrays.asList(new PHPTokenId[]{
+                PHPTokenId.PHP_PRIVATE,
+                PHPTokenId.PHP_PUBLIC,
+                PHPTokenId.PHP_PROTECTED,
+                PHPTokenId.PHP_ABSTRACT,
+                PHPTokenId.PHP_VAR,
+                PHPTokenId.PHP_STATIC,
+                PHPTokenId.PHP_CONST
+            }));
+        }
+        return offerMagicAndInherited;
+    }
+
+    private boolean completeFieldTypes(TokenSequence<PHPTokenId> tokenSequence, int caretOffset, TokenHierarchy<?> th, FileObject fileObject) {
+        if (!isPhp74OrNewer(fileObject)) {
```

Comment:
I am not sure whether this will not cause any performance problems... We can wait, maybe I am wrong.

Reply:
Does it mean that we should not check php version?

Reply:
Let's keep it as it is but cc is called very often and getting PHP version for a file means an IO operation. Yes, SSD disks are common today so maybe it won't be a problem. We will possibly see, in the profiler snapshots :)

Reply:
Makes sense.